### DISCUSSION TITLE
How to use fixed pin location file in ORFS flow?

### DISCUSSION THREAD

**Discussion by vijayank88** [The-OpenROAD-Project/OpenROAD-flow-scripts#422] [Q&A] [answered]
This question came from gitter channel:
Is it possible to manually specify the pins locations instead of having a complete random placement.
I do see the following directive used in https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/blob/master/flow/designs/gf12/swerv_wrapper/config.mk
`export PLACE_PINS_ARGS = -exclude left: -exclude right: -exclude top:* -exclude bottom:0-10 -exclude bottom:400-700`
Can I use `PLACE_PINS_ARGS` to specify the location of the pins? If yes what is the syntax to be used?

Given that I essentially have pins with fix locations, through which variable in the `config.mk`  file should I pass the name of the file containing the `place_pin` set of directives? Would something like `IO_CONSTRAINTS=my_place_pin_directives.tcl`  be a good choice? If not which other variable should I use in this case?

**Comment by vvbandeira** [comment]
I don't think there's a straightforward way to do this by just modifying existing flow arguments. Using `IO_CONSTRAINTS` would probably not get the desired result because we call `place_pins` after sourcing the file, see [here](https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/blob/ed462b5dcab174fc861b954a8d4965ba1be2e45e/flow/scripts/io_placement.tcl#L4-L9). The user can replace the default ORFS `io_placement.tcl` script with their own version. The user can also file an Issue requesting the feature or submit a PR.

**Comment by maliberty** [comment]
You can use place_pin (see https://github.com/The-OpenROAD-Project/OpenROAD/tree/master/src/ppl) for manual placement.

**Reply by gkamendje** [reply]
@maliberty 
Assuming I have all my 20 pins that I need to place on a specific location. In which file should I put the 20 place_pin directives? I don't think that modifying the default ORFS `io_placement.tcl` is the most elegant way to achieve this. Maybe having an option for the user to supply a file with pin placement directives would be more appropriate.

**Comment by maliberty** [comment]
In io_placement.tcl you can use

if {[info exists ::env(IO_CONSTRAINTS)]} {
  source $::env(IO_CONSTRAINTS)
}

to give your placement.

**Reply by gkamendje** [reply]
@maliberty 
vvbandeira was suggesting that using IO_CONSTRAINTS this way might not be the best solution given that `place_pins` is called right after sourcing the content of `IO_CONSTRAINTS` in `io_placement.tcl`. Do you suggest to comment out the call to `place_pins` in this case?

**Reply by maliberty** [reply]
I believe it will be a no-op if you have manually placed all the pins (or at least it should be) so there should be no harm in running it.

**Comment by laurentc2** [comment]
Is there somewhere an example using **place_pin** that set/force the all the pins name, location and layer ?
Such an example for that function would be nice .

Laurent

**Reply by maliberty** [reply]
I don't know that there is an example in ORFS but you can see the unit test https://github.com/The-OpenROAD-Project/OpenROAD/blob/master/src/ppl/test/place_pin1.tcl

**Comment by vijayank88** [comment, marked as answer]
With `set_io_pin_constraint` you can restrict IO placement on specific side. 
Examples found here: https://github.com/The-OpenROAD-Project/OpenROAD/blob/master/src/ppl/test/add_constraint10.tcl
